qcodes/instrument_drivers/QuantumDesign/DynaCoolPPMS/DynaCool.py

```diff
@@ -0,0 +1,146 @@
+                           unit='K',
+                           get_parser=partial(DynaCool._pick_one, 1, float),
+                           get_cmd='CHAT?')
+
+        # The error code of the latest command
+        self._error_code = 0
+
+        self._update_temperatures()
+
+        self.connect_message()
+
+    @property
+    def error_code(self):
+        return self._error_code
+
+    @staticmethod
+    def _pick_one(which_one: int, parser: type, resp: str) -> str:
+        """
+        Since most of the API calls return several values in a comma-seperated
+        string, here's a convenience function to pick out the substring of
+        interest
+        """
+        return parser(resp.split(', ')[which_one])
+
+    def get_idn(self) -> Dict[str, Optional[str]]:
+        response = self.ask('*IDN?')
+        # just clip out the error code
+        idparts = response[2:].split(', ')
+
+        return dict(zip(('vendor', 'model', 'serial', 'firmware'), idparts))
+
+    def _update_temperatures(self) -> None:
+        """
+        This function queries the last temperature setpoint (w. rate and mode)
+        from the instrument.
+        """
+        raw_response = self.ask('GLTS?')
+        sp = DynaCool._pick_one(1, float, raw_response)
+        rate = DynaCool._pick_one(2, float, raw_response)
+        mode = DynaCool._pick_one(3, int, raw_response)
+
+        self.temperature_setpoint._save_val(sp)
+        self.temperature_rate.raw_value = rate
+        self.temperature_rate._save_val(rate/60)
+        self.temperature_settling.raw_value = mode
+        inv_map = {val: key for (key, val) in
+                   self.temperature_settling.val_mapping.items()}
+        self.temperature_settling._save_val(inv_map[mode])
+
+    def _temp_setter(self, param: str, value: float) -> None:
+        self._update_temperatures()
+        params = ['temperature_setpoint', 'temperature_rate',
+                  'temperature_settling']
+        vals = list(self.parameters[par].raw_value for par in params)
+        vals[params.index(param)] = value
+
+        self.write(f'TEMP {vals[0]}, {vals[1]}, {vals[2]}')
+
+    def write(self, cmd: str) -> None:
+        """
+        Since the error code is always returned, we must read it back
+        """
+        super().write(cmd)
+        self._error_code = self.visa_handle.read()
+
+    def ask(self, cmd: str) -> str:
+        """
+        Since the error code is always returned, we must read it back
+        """
+        response = super().ask(cmd)
+        self._error_code = DynaCool._pick_one(0, str, response)
+        return response
+
+    def close(self) -> None:
+        """
+        Make sure to nicely close the server connection
+        """
+        try:
+            log.debug('Closing server connection.')
+            self.write('CLOSE')
+        except VisaIOError:
+            log.info('Could not close connection to server, perhaps the '
```

Comment:
could you also log the exception itself so that debugging with logging is easier?

Reply:
Done it now, but this try-except is just to allow users to close the server *before* closing the client session without getting upsetting errors.